### PR TITLE
feat(auth): add option to define timeout

### PR DESCRIPTION
**Pull request by stonebuzz**
Add new option from authldap (and authldapreplicate) form to define timeout 

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #number

### CODE REVIEW

File: install/migrations/update_9.5.x_to_10.0.0/ldaps.php [hunk 1]
```diff
@@ -71,3 +71,27 @@
...
 }
+
+if (!$DB->fieldExists('glpi_authldaps', 'timeout')) {
+   $migration->addField(
+      'glpi_authldaps',
+      'timeout',
+      'int',
+      [
+         'after'  => 'use_bind',
+         'value' => 0
```
Comment (flegastelois) [inline install/migrations/update_9.5.x_to_10.0.0/ldaps.php:82]: maybe "3" per default ?

File: install/migrations/update_9.5.x_to_10.0.0/ldaps.php [hunk 2]
```diff
@@ -71,3 +71,27 @@
...
+}
+
+if (!$DB->fieldExists('glpi_authldapreplicates', 'timeout')) {
+   $migration->addField(
+      'glpi_authldapreplicates',
+      'timeout',
+      'int',
+      [
+         'after'  => 'name',
+         'value' => 0
```
Comment (flegastelois) [inline install/migrations/update_9.5.x_to_10.0.0/ldaps.php:94]: maybe "3" per default ?

Comment (stonebuzz) [reply install/migrations/update_9.5.x_to_10.0.0/ldaps.php]: done

File: install/migrations/update_9.5.x_to_10.0.0/ldaps.php (hunk 1 above)
Comment (stonebuzz) [reply install/migrations/update_9.5.x_to_10.0.0/ldaps.php]: done

File: inc/authldap.class.php [hunk 3]
```diff
@@ -403,6 +403,11 @@ function showForm($ID, $options = []) {
          Dropdown::showYesNo('use_bind', $this->fields["use_bind"]);
          echo "</td></tr>";
 
+         echo "<tr class='tab_bg_1'><td><label for='timeout'>" . __('Timeout') . "</label></td>";
+         echo "<td colspan='3'>";
+         Dropdown::showNumber('timeout', ['value' => $this->fields["timeout"] , 'min' => 3, 'max' => 10, 'step' => 1]);
```
Comment (cedric-anne) [inline inc/authldap.class.php:408]: I do not know the average response time of LDAP servers, but I guess administrators may want to use a `1` second timeout (LDAP has replicates, and we use them when main server does not respond instantly), or a `30` seconds timeout (LDAP has no replicates, and network connection between GLPI and LDAP server is not so stable).

```suggestion
         Dropdown::showNumber('timeout', ['value' => $this->fields["timeout"] , 'min' => 1, 'max' => 30, 'step' => 1]);
```

Comment (flegastelois) [inline inc/authldap.class.php:408]:
```suggestion
         Dropdown::showNumber('timeout', ['value' => $this->fields["timeout"] , 'min' => 0, 'max' => 10, 'step' => 1]);
```

File: inc/authldapreplicate.class.php [hunk 4]
```diff
@@ -79,14 +79,18 @@ static function addNewReplicateForm($target, $master_id) {
...
       echo "<td class='center'>".__('Server')."</td>";
-      echo "<td class='center'>"._n('Port', 'Ports', 1)."</td><td></td></tr>";
+      echo "<td class='center'>"._n('Port', 'Ports', 1)."</td>";
+      echo "<td class='center'>".__('Timeout')."</td><td></td></tr>";
       echo "<tr class='tab_bg_1'>";
       echo "<td class='center'><input type='text' name='name'></td>";
       echo "<td class='center'><input type='text' name='host'></td>";
       echo "<td class='center'><input type='text' name='port'></td>";
+      echo "<td class='center'>";
+      Dropdown::showNumber('timeoout', ['min' => 3, 'max' => 10, 'step' => 1]);
```
Comment (flegastelois) [inline inc/authldapreplicate.class.php:92]:
```suggestion
      Dropdown::showNumber('timeoout', ['min' => 0, 'max' => 10, 'step' => 1]);
```

**Review by flegastelois — APPROVED** [review]

File: inc/authldap.class.php [hunk 5]
```diff
@@ -1194,6 +1201,16 @@ function rawSearchOptions() {
          'autocomplete'       => true,
       ];
 
+      $tab[] = [
+         'id'                 => '32',
+         'table'              => $this->getTable(),
+         'field'              => 'timeout',
+         'name'               => __('Timeout'),
+         'massiveaction'      => false,
+         'datatype'           => 'text',
```
Comment (cedric-anne) [inline inc/authldap.class.php:1210]:
```suggestion
         'datatype'           => 'number',
         'unit'                   => 'second',
```

File: inc/authldapreplicate.class.php (hunk 4 above)
Comment (cedric-anne) [inline inc/authldapreplicate.class.php:92]:
```suggestion
      Dropdown::showNumber('timeout', ['min' => 3, 'max' => 10, 'step' => 1]);
```

File: inc/authldap.class.php [hunk 6]
```diff
@@ -2659,13 +2678,14 @@ function connect() {
...
                                    $tls_certfile = "", $tls_keyfile = "",
-                                   $use_bind = true) {
+                                   $use_bind = true, $timeout = 0) {
 
       $ds = @ldap_connect($host, intval($port));
       if ($ds) {
          @ldap_set_option($ds, LDAP_OPT_PROTOCOL_VERSION, 3);
          @ldap_set_option($ds, LDAP_OPT_REFERRALS, 0);
          @ldap_set_option($ds, LDAP_OPT_DEREF, $deref_options);
+         @ldap_set_option($ds, LDAP_OPT_NETWORK_TIMEOUT, $timeout);
```
Comment (cedric-anne) [inline inc/authldap.class.php:2688]: Maybe this option should not be set if timeout is 0.
```suggestion
         if ($timeout > 0) {
            @ldap_set_option($ds, LDAP_OPT_NETWORK_TIMEOUT, $timeout);
         }
```

File: inc/authldap.class.php [hunk 7]
```diff
@@ -1194,6 +1201,16 @@ function rawSearchOptions() {
          'autocomplete'       => true,
       ];
 
+      $tab[] = [
+         'id'                 => '32',
+         'table'              => $this->getTable(),
+         'field'              => 'timeout',
+         'name'               => __('Timeout'),
+         'massiveaction'      => false,
+         'datatype'           => 'number',
```
Comment (cedric-anne) [inline inc/authldap.class.php:1210]:
```suggestion
         'datatype'           => 'number',
         'unit'               => 'second',
```

Comment (cedric-anne) [reply inc/authldap.class.php:1210]: Not sure about that. If we allow `0` value, it should not display "0 second".

File: inc/authldap.class.php (hunk 3 above)
Comment (stonebuzz) [reply inc/authldap.class.php]: done

File: inc/authldapreplicate.class.php (hunk 4 above)
Comment (stonebuzz) [reply inc/authldapreplicate.class.php]: done

Comment (stonebuzz) [reply inc/authldapreplicate.class.php]: done

File: inc/authldap.class.php (hunk 7 above)
Comment (stonebuzz) [reply inc/authldap.class.php:1215]: done

File: inc/authldap.class.php (hunk 3 above)
Comment (stonebuzz) [reply inc/authldap.class.php]: done

File: inc/authldap.class.php [hunk 8]
```diff
@@ -1194,6 +1206,20 @@ function rawSearchOptions() {
...
+         'unit'               => 'second',
+         'toadd'              => [
+            '0'                  => 'No timeout'
```
Comment (cedric-anne) [inline inc/authldap.class.php:1222]:
```suggestion
            '0'                  => __('No timeout')
```

File: inc/authldap.class.php [hunk 9]
```diff
@@ -403,6 +403,20 @@ function showForm($ID, $options = []) {
          Dropdown::showYesNo('use_bind', $this->fields["use_bind"]);
          echo "</td></tr>";
 
+         echo "<tr class='tab_bg_1'><td><label for='timeout'>" . __('Timeout') . "</label></td>";
+         echo "<td colspan='3'>";
+
+         //set deffault value to 10 on create
+         if (empty($this->fields["timeout"])) {
+            $this->fields["timeout"] = 10;
+         }
```
Comment (cedric-anne) [inline inc/authldap.class.php:412]:
```suggestion
         if ($this->isNewItem()) {
            $this->fields["timeout"] = 10;
         }
```

File: inc/authldap.class.php [hunk 10]
```diff
@@ -403,6 +403,20 @@ function showForm($ID, $options = []) {
...
+         Dropdown::showNumber('timeout', ['value'  => $this->fields["timeout"],
+                                          'min'    => 1,
+                                          'max'    => 10,
```
Comment (cedric-anne) [inline inc/authldap.class.php:415]:
```suggestion
                                          'max'    => 30,
```

File: inc/authldap.class.php (hunk 9 above)
Comment (stonebuzz) [reply inc/authldap.class.php]: Done

File: inc/authldap.class.php (hunk 8 above)
Comment (stonebuzz) [reply inc/authldap.class.php]: Done

File: inc/authldap.class.php (hunk 10 above)
Comment (stonebuzz) [reply inc/authldap.class.php]: Done

**Review by cedric-anne — APPROVED** [review]

**Review by trasher — APPROVED** [review]

File: inc/authldap.class.php [hunk 11]
```diff
@@ -562,6 +548,21 @@ function showFormAdvancedConfig() {
...
       echo "</tr>";
 
+
+      echo "<tr class='tab_bg_1'><td><label for='timeout'>" . __('Timeout') . "</label></td>";
+      echo "<td colspan='3'>";
+
+      //set deffault value to 10 on create
+      if ($this->isNewItem()) {
+         $this->fields["timeout"] = 10;
+      }
```
Comment (cedric-anne) [inline inc/authldap.class.php:557]: This is now useless. Creation will use the default value from DB schema, which is 10.
```suggestion
```

**Review by orthagh — APPROVED** [review]

File: inc/authldap.class.php (hunk 11 above)
Comment (stonebuzz) [reply inc/authldap.class.php]: done

**Review by cedric-anne — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Pull request merged**